aten/src/ATen/SparseTensorImpl.h

```diff
@@ -210,6 +210,32 @@ struct CAFFE2_API SparseTensorImpl : public TensorImpl {
     impl->refresh_numel();
     return impl;
   }
+
+  // NOTE: `shallow_copy_from()` does not copy the AutogradMeta pointer
+  // because it is unique for each Variable.
+  virtual void shallow_copy_from(c10::intrusive_ptr<TensorImpl> impl) override {
+    AT_ASSERT(impl->is_sparse());
+    auto sparse_impl = static_cast<SparseTensorImpl*>(impl.get());
+    type_id_ = sparse_impl->type_id();
+    data_type_ = sparse_impl->dtype();
+
+    // TensorImpl general fields
+    // Note that these fields are not used in sparse tensor code, and we copy them here only for completeness.
+    sizes_ = sparse_impl->sizes_;
+    strides_ = sparse_impl->strides_;
+    storage_offset_ = sparse_impl->storage_offset_;
+    is_contiguous_ = sparse_impl->is_contiguous_;
+    is_wrapped_number_ = sparse_impl->is_wrapped_number_;
+    reserved_ = sparse_impl->reserved_;
+
+    // Sparse-specific fields
+    sparse_dim_ = sparse_impl->sparse_dim();
+    dense_dim_ = sparse_impl->dense_dim();
+    indices_ = sparse_impl->indices();
+    values_ = sparse_impl->values();
+    coalesced_ = sparse_impl->coalesced();
+    refresh_numel();
+  }
```

Comment:
This method (and the one in TensorImpl.h) is necessary for `variable.set_data(data_tensor)` to work, because the new implementation of `variable.set_data(data_tensor)` is to shallow-copy all TensorImpl fields (except AutogradMeta pointer) from `data_tensor` to `variable`, so that the semantics of `set_data()` can be preserved (i.e. setting the data of a Variable without changing its autograd fields).

Reply:
Update: Added `NOTE [ TensorImpl Shallow-Copying ]` to TensorImpl.h to explain this.